src/android/ManagerService.java

```diff
@@ -0,0 +1,470 @@
+
+        if (!isNetworkAvailable) {
+            logMessage(String.format("eventLabel='Uploader no network available, upload has been queued' uploadId='%s'", uploadId));
+            return;
+        }
+
+        MultipartUploadRequest request;
+
+        try {
+            request = new MultipartUploadRequest(this, payload.get("serverUrl").toString())
+                    .setUploadID(uploadId)
+                    .setMethod("POST")
+                    .addFileToUpload(payload.get("filePath").toString(), payload.get("fileKey").toString())
+                    .setMaxRetries(0);
+        } catch (IllegalArgumentException | FileNotFoundException error) {
+            sendAddingUploadError(uploadId, error);
+            return;
+        }
+
+        try {
+            HashMap<String, Object> headers = convertToHashMap((JSONObject) payload.get("headers"));
+            for (String key : headers.keySet()) {
+                request.addHeader(key, headers.get(key).toString());
+            }
+        } catch (JSONException exception) {
+            logMessage(String.format("eventLabel='could not parse request headers' uploadId='%s' error='%s'", uploadId, exception.getMessage()));
+            sendAddingUploadError(uploadId, exception);
+            return;
+        }
+
+        try {
+            HashMap<String, Object> parameters = convertToHashMap((JSONObject) payload.get("parameters"));
+            for (String key : parameters.keySet()) {
+                request.addParameter(key, parameters.get(key).toString());
+            }
+        } catch (JSONException exception) {
+            logMessage(String.format("eventLabel='could not parse request parameters' uploadId='%s' error='%s'", uploadId, exception.getMessage()));
+            sendAddingUploadError(uploadId, exception);
+            return;
+        }
+
+        request.startUpload();
+    }
+
+    private void sendAddingUploadError(String uploadId, Exception error) {
+        deletePendingUploadAndSendEvent(new JSONObject(new HashMap() {{
+            put("id", uploadId);
+            put("state", "FAILED");
+            put("errorCode", 0);
+            put("error", error.getMessage());
+        }}));
+    }
+
+    public static HashMap<String, Object> convertToHashMap(JSONObject jsonObject) throws JSONException {
+        HashMap<String, Object> hashMap = new HashMap<>();
+        if (jsonObject != null) {
+            Iterator<?> keys = jsonObject.keys();
+            while (keys.hasNext()) {
+                String key = (String) keys.next();
+                Object value = jsonObject.get(key);
+                hashMap.put(key, value);
+            }
+        }
+        return hashMap;
+    }
+
+    private void sendMissingEvents() {
+        migrateOldUploads();
+
+        for (UploadEvent event : UploadEvent.all()) {
+            logMessage("Uploader send event missing on Start - " + event.getId());
+            this.connectedPlugin.callback(event.dataRepresentation());
```

Comment:
Ici je pense que c'est plutot:

```suggestion
            this.sendCallback(event.dataRepresentation());
```

Et est-ce que on peut changer la logique pour faire quelque chose comme ca:

```
while (event = UploadEvent.first())
  this.sendCallback(event.dataRepresentation());
  sleep(250ms)
```

Il faut verifier aussi que l'appel n'est pas blocant... ou faire ca dans un autre thread.

Reply:
done